lib/router/route-info.ts

```diff
@@ -37,29 +37,40 @@ export type Continuation = () => PromiseLike<boolean> | boolean;
 
 export interface RouteInfo {
   readonly name: string;
-  readonly parent: Maybe<RouteInfo>;
-  readonly child: Maybe<RouteInfo>;
+  readonly parent: RouteInfo | null;
+  readonly child: RouteInfo | null;
   readonly localName: string;
   readonly params: Dict<unknown>;
+  readonly queryParams: Dict<unknown>;
   find(
-    predicate: (this: void, routeInfo: RouteInfo, i: number) => boolean,
-    thisArg: any
+    predicate: (this: any, routeInfo: RouteInfo, i?: number) => boolean,
+    thisArg?: any
   ): RouteInfo | undefined;
 }
 
-let ROUTE_INFO_LINKS = new WeakMap<InternalRouteInfo<Route>, RouteInfo>();
+let INTERNAL_TO_PUBLIC = new WeakMap<InternalRouteInfo<Route>, RouteInfo>();
 
-export function toReadOnlyRouteInfo(routeInfos: InternalRouteInfo<Route>[]) {
+export function toReadOnlyRouteInfo(
+  routeInfos: InternalRouteInfo<Route>[],
+  queryParams: Dict<unknown> = {}
+) {
   return routeInfos.map((info, i) => {
-    let { name, params, queryParams, paramNames } = info;
+    let { name, params, paramNames } = info;
     let publicRouteInfo = new class implements RouteInfo {
```

Comment:
Is there a particular reason this class has to be generated each time `toReadOnlyRouteInfo` is called?

Reply:
One would think you could use the `WeakMap` as a cache, however there seems to be some mutation going on related to QPs which results in a corrupted cache.

Reply:
After further investigation, this does not work without significant work around mutability of ResolvedRouteInfos.